Packages/Tracking Preview/PhysicsHands/Runtime/UI/Scripts/PhysicsButton.cs

```diff
@@ -0,0 +1,506 @@
+            {
+                _isPressed = false;
+                _unpressedThisFrame = true;
+                OnUnpress?.Invoke();
+            }
+
+            pressedAmount = Mathf.InverseLerp(buttonHeightLimit * 0.99f, buttonHeightLimit * 0.01f, _buttonElement.transform.localPosition.y);
+
+            if (_pressedThisFrame && isToggleable)
+            {
+                _isToggled = !_isToggled;
+                UpdateToggleJoint();
+                OnToggle?.Invoke(_isToggled);
+            }
+        }
+
+        private void PressValidation()
+        {
+            if (!_isPressed && _buttonElement.transform.localPosition.y <= buttonHeightLimit * 0.01f)
+            {
+                _isPressed = true;
+                _pressedThisFrame = true;
+                OnPress?.Invoke();
+            }
+        }
+
+        private void LateUpdate()
+        {
+            if (_pressedThisFrame)
+            {
+                _pressedThisFrame = false;
+            }
+            if (_unpressedThisFrame)
+            {
+                _unpressedThisFrame = false;
+            }
+        }
+
+        private void SetupButton()
+        {
+            if (_buttonElement == null)
+                return;
+
+            _rigidbody.mass = 1f;
+            _rigidbody.isKinematic = true;
+            _rigidbody.useGravity = false;
+            _rigidbody.collisionDetectionMode = CollisionDetectionMode.ContinuousSpeculative;
+
+            _buttonElement.Rigid.mass = 1f;
+            _buttonElement.Rigid.drag = 0;
+            _buttonElement.Rigid.useGravity = false;
+            _buttonElement.Rigid.isKinematic = false;
+            _buttonElement.Rigid.collisionDetectionMode = CollisionDetectionMode.ContinuousSpeculative;
+
+            SetupJoint(_buttonElement.Joint);
+        }
+
+        private void SetupJoint(ConfigurableJoint joint)
+        {
+            joint.connectedBody = _rigidbody;
+            joint.anchor = Vector3.up * 0.5f;
+            joint.axis = Vector3.right;
+            joint.autoConfigureConnectedAnchor = false;
+            joint.connectedAnchor = Vector3.up * buttonHeightLimit;
+
+            joint.xMotion = ConfigurableJointMotion.Locked;
+            joint.yMotion = ConfigurableJointMotion.Limited;
+            joint.zMotion = ConfigurableJointMotion.Locked;
+            joint.angularXMotion = ConfigurableJointMotion.Locked;
+            joint.angularYMotion = ConfigurableJointMotion.Locked;
+            joint.angularZMotion = ConfigurableJointMotion.Locked;
+
+            SoftJointLimit softLimit = joint.linearLimit;
+            softLimit.limit = buttonHeightLimit / 2f;
+            softLimit.bounciness = 0f;
+            softLimit.contactDistance = 0f;
+            joint.linearLimit = softLimit;
+
+            joint.xDrive = new JointDrive();
+            joint.zDrive = new JointDrive();
+
+            JointDrive yDrive = joint.yDrive;
```

Comment:
What's the reason for taking the local copy here - to set all three as an atomic operation?

Reply:
X and Z drives need to be absolutely nothing but the Y drive is the one we use for the button. This ensures that if the user makes changes to the buttons drives, they are restored back to the defaults.